core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -257,10 +273,48 @@ public Snapshot apply() {
           .run(index -> manifestFiles[index] = manifestsWithMetadata.get(manifests.get(index)));
 
       writer.addAll(Arrays.asList(manifestFiles));
+
+      if (manifestListKeyMetadata != null) {
+        writer.close();
+        manifestListSize = writer.length();
```

Comment:
For manifests, we use a pattern where the writer produces the `ManifestFile` metadata after closing the writer to delegate more responsibility to the writer and make it easy to produce the metadata. I'm wondering if that's a good idea here as well, since the writer will have the `EncryptedOutputFile`.

The main difference for the pattern is that the key metadata from the `EncryptedOutputFile` is unencrypted and needs to be encrypted using the key encryption key. I don't think I have a strong preference whether this is delegated to the writer (`toManifestListFile()`) or just done in a method like `ManifestListFile.create(EncryptionManager em, String wrappedKey, ByteBuffer keyMetadataBuffer, String location)`. Either way, I don't think that it should be done here.